<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>exercise-01.knit</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="schedule.html">Schedule</a>
</li>
<li>
  <a href="resources.html">Resources</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<p><br />
</p>
<div
id="bayesian-tree-inference-using-the-fossilised-birth-death-process-in-revbayes"
class="section level1">
<h1>Bayesian tree inference using the fossilised birth-death process in
RevBayes</h1>
<div id="contents" class="section level4">
<h4>Contents</h4>
<ul>
<li><a href="#start">Getting started</a></li>
<li><a href="#read">Read the data</a></li>
<li><a href="#time">Time tree model</a></li>
<li><a href="#clock">Strict clock model</a></li>
<li><a href="#Mk">Substitution model</a></li>
<li><a href="#mcmc">MCMC settings</a></li>
<li><a href="#output">Evaluating the output</a></li>
<li><a href="#next">Next tasks</a></li>
</ul>
<p>In this exercise we’ll estimate a time tree of bears under the
fossilised birth-death (FBD) process. We’ll use a <a
href="data/bears_morphology.nex">morphological matrix</a> that comprises
62 binary characters for 18 living and fossil bears. We’ll also use <a
href="data/bears_taxa.tsv">age information</a> associated with our taxa.
Download these files and examine the contents. The exercise is adapted
from this <a href="https://hal.inria.fr/PGE/hal-02536394">published
version</a> of the exercise, which contains more background
information.</p>
</div>
<div id="software" class="section level4">
<h4>Software</h4>
<p>Download the software <a
href="https://revbayes.github.io/download">RevBayes</a>, <a
href="https://beast.community/tracer">Tracer</a> and <a
href="https://beast.community/figtree">FigTree</a> if you haven’t done
so already. This tutorial uses RevBayes 1.2.3.</p>
<p>Edit the scripts in a text editor of your choice.</p>
</div>
<div id="start" class="section level2">
<h2>Getting started</h2>
<p>For this exercise create a folder called
<code>FBD_exercise_RevBayes</code> (or whatever you want). Then create
two sub-directories: <code>data</code> and <code>scripts</code>.
Download the data and move it to the <code>data</code> directory.</p>
<p>Recall that for divergence time estimation we need three model
components, so we’ll create the following scripts:</p>
<ol style="list-style-type: decimal">
<li><code>main.Rev</code> for reading the data, organising calls to
other scripts and the MCMC settings</li>
<li><code>scripts/FBD_dating.Rev</code> for the time tree model</li>
<li><code>scripts/Mk+G.Rev</code> for the substitution model</li>
<li><code>scripts/clock_strict_morpho.Rev</code> for the clock
model</li>
</ol>
</div>
<div id="read" class="section level2">
<h2>Read in the data and define helper variables</h2>
<p>Start with your <code>main.Rev</code> script. As in previous RevBayes
exercises, we’ll start by reading in the data and defining some helper
variables.</p>
<pre class="r"><code>taxa &lt;- readTaxonData(&quot;data/bears_taxa.tsv&quot;)
morpho &lt;- readDiscreteCharacterData(&quot;data/bears_morphology.nex&quot;)</code></pre>
<pre class="r"><code>num_taxa &lt;- morpho.ntaxa() # number of taxa
num_branches &lt;- 2 * num_taxa - 2 # number of branches in an rooted tree</code></pre>
<p>Define a set of variables for setting up our MCMC.</p>
<pre class="r"><code>moves    = VectorMoves()
monitors = VectorMonitors()</code></pre>
</div>
<div id="time" class="section level2">
<h2>Time tree model</h2>
<p>We will use the FBD process as our time tree model. This describes
the distribution of most probable trees, given the speciation,
extinction, and fossil sampling processes, combined with the fossil age
information.</p>
<p>Go to your <code>FBD_dating.Rev</code> script to set this up.</p>
<p>Three of the key parameters of the FBD model are speciation (or
birth) rate (<span class="math inline">\(\lambda\)</span>), extinction
(or death) rate (<span class="math inline">\(\mu\)</span>) and fossil
sampling rate (<span class="math inline">\(\psi\)</span>). We will
assume these rates are constant over time and place exponential priors
on each of these.</p>
<pre class="r"><code>speciation_rate ~ dnExponential(10)
extinction_rate ~ dnExponential(10)
fossil_samp_rate ~ dnExponential(10)</code></pre>
<p>Each parameter is a stochastic variable drawn from an exponential
distribution with <span class="math inline">\(\delta\)</span> = 10 and a
mean of <span class="math inline">\(1/\delta\)</span> = 0.1. Estimates
of speciation, extinction and sampling tend to be on this order.</p>
<p>Next specify the moves on these parameters.</p>
<pre class="r"><code>moves.append( mvScale(speciation_rate, lambda = 0.5, tune = true, weight = 3.0) )
moves.append( mvScale(extinction_rate, lambda = 0.5, tune = true, weight = 3.0) )
moves.append( mvScale(fossil_samp_rate, lambda = 0.5, tune = true, weight = 3.0) )</code></pre>
<p>We might also be interested in diversification and turnover, so let’s
set up a deterministic variable for each of these.</p>
<pre class="r"><code>diversification := speciation_rate - extinction_rate
turnover := extinction_rate/speciation_rate</code></pre>
<p>Next we’ll define the probability of extant species sampling (<span
class="math inline">\(\rho\)</span>). Since we sample all extant bears,
we’ll specify this probability as a constant variable = 1.0.</p>
<pre class="r"><code>rho &lt;- 1.0</code></pre>
<p>Now we’ll specify a prior on the origin time parameter. The origin
represents the beginning of the process and will be older than the root.
It’s a bit tricky to interpret this parameter, but you can think of it
as the branching time between bears and the broader tree of mammals.
We’ll use the oldest bear fossil as a minimum (35.0) and a recent
estimate for the age of carnivores as a maximum (55.0).</p>
<pre class="r"><code># prior and move on the origin
origin_time ~ dnUnif(37.0, 55.0)
moves.append( mvSlide(origin_time, weight = 1.0) )</code></pre>
<p>Next we’ll specify the FBD model, along with moves on the tree
topology and branch lengths.</p>
<pre class="r"><code>tree ~ dnBDSTP(lambda = speciation_rate, mu = extinction_rate, psi = fossil_samp_rate, rho = rho, origin = origin_time, taxa = taxa, r = 0)

# moves to search tree space
moves.append( mvFNPR(tree, weight = 15.0) )
moves.append( mvCollapseExpandFossilBranch(tree, origin_time, weight = 6.0) )
moves.append( mvNodeTimeSlideUniform(tree, weight = 40.0) )
moves.append( mvRootTimeSlideUniform(tree, origin_time, weight = 5.0) )
moves.append( mvFossilTimeSlideUniform(tree, origin_time, weight = 5.0) )</code></pre>
<p>We can create “clade constraints” to monitor the age of specific
clades of interest. Let’s do that for the clade containing all living
bears.</p>
<pre class="r"><code>clade_extant = clade(&quot;Ailuropoda_melanoleuca&quot;,&quot;Tremarctos_ornatus&quot;,&quot;Melursus_ursinus&quot;,
                    &quot;Ursus_arctos&quot;,&quot;Ursus_maritimus&quot;,&quot;Helarctos_malayanus&quot;,
                    &quot;Ursus_americanus&quot;,&quot;Ursus_thibetanus&quot;)

age_extant := tmrca(tree, clade_extant)</code></pre>
<p>We might also be interested in know how many sampled ancestors are in
our posterior trees, so let’s set up another deterministic variable to
monitor this.</p>
<pre class="r"><code>num_samp_anc := tree.numSampledAncestors()</code></pre>
</div>
<div id="clock" class="section level2">
<h2>Strict clock model</h2>
<p>For simplicity, we’ll use a strict clock model to describe the rate
of character change, assuming the rate is constant over time and across
all branches in our tree.</p>
<p>Open your <code>clock_strict_morpho.Rev</code> script. We don’t know
the rate of evolution, so as before we’ll use an exponential prior, with
a mean of 0.01, since rates of morphological evolution tend to be
low.</p>
<pre class="r"><code>branch_rates ~ dnExponential(100)

moves.append( mvScale(branch_rates, lambda = 0.5, tune = true, weight = 3.0) )</code></pre>
<p>This rate will be used for all branches.</p>
</div>
<div id="Mk" class="section level2">
<h2>Substitution model</h2>
<p>Next we’ll specify the Mk substitution model in the
<code>Mk+G.Rev</code> file, which describes the probability of
transitioning from one character state to another. This model is a
generalisation of the Jukes-Cantor model for <span
class="math inline">\(k\)</span> states.</p>
<p>First we’ll use the <code>fnJC</code> function to define a Q matrix
for 2 states (0, 1), since we only have two character states in our
morphological character matrix. Remember that under this model rates of
change between states and state frequencies are equal.</p>
<pre class="r"><code>Q &lt;- fnJC(2)</code></pre>
<p>Then we’ll add the following lines to add the Gamma sites model,
using an exponential prior with mean = 1 on the shape parameter
<code>alpha</code> and 4 discrete rate categories.</p>
<pre class="r"><code>alpha_morpho ~ dnExponential( 1.0 )
rates_morpho := fnDiscretizeGamma( alpha_morpho, alpha_morpho, 4 )

moves.append( mvScale(alpha_morpho, lambda=0.1, tune = true, weight=3.0) )</code></pre>
<p>Then we’ll define a stochastic node representing our character matrix
and “clamp” that variable to our morphological data. Note that this time
we need to use <code>type = "Standard"</code>. I’m not sure why RevBayes
uses the term “standard” for morphological data.</p>
<pre class="r"><code>seq ~ dnPhyloCTMC(tree = tree, Q = Q, type = &quot;Standard&quot;, branchRates = branch_rates, siteRates = rates_morpho)
seq.clamp(morpho)</code></pre>
</div>
<div id="mcmc" class="section level2">
<h2>MCMC settings</h2>
<p>Back in <code>main.Rev</code> add the following lines to include the
tree, clock and substitution models.</p>
<pre class="r"><code># tree model
source(&quot;scripts/FBD_dating.Rev&quot;)
# clock model
source(&quot;scripts/clock_strict_morpho.Rev&quot;)
# substitution model
source(&quot;scripts/Mk+G.Rev&quot;)</code></pre>
<p>First we create the variable <code>mymodel</code>.</p>
<pre class="r"><code>mymodel = model(tree)</code></pre>
<p>Then we define a set of monitors to capture the output.</p>
<pre class="r"><code># parameters printed to file
monitors.append( mnModel(filename = &quot;output/bears_FBD_morpho.log&quot;, printgen = 100) )
# trees printed to file
monitors.append( mnFile(filename = &quot;output/bears_FBD_morpho.trees&quot;, printgen = 100, tree) )
# parameter values printed to screen during the MCMC
monitors.append( mnScreen(printgen = 100, age_extant, num_samp_anc, origin_time) )</code></pre>
<p>Finally, we’ll set up the MCMC run using the <code>mcmc</code>
function, specifying our model, the vector of monitors and the vector of
moves and run 2 chains for 100000 generations.</p>
<pre class="r"><code>mymcmc = mcmc(mymodel, monitors, moves, nruns = 2, combine = &quot;mixed&quot;)
mymcmc.run(generations = 100000, tuningInterval = 1000)</code></pre>
<p>Note we’ve added the argument <code>tuningInterval = 1000</code>.
This tells the chain how long to spend optimising the MCMC moves.
Different datasets behave differently <span
class="math inline">\(-\)</span> the above option partly automates the
selection of parameters that maximise efficiency of the MCMC but it can
be more effective to apply multiple moves to the same parameter.</p>
<p>Add the following to the end of your script to generate a summary
tree.</p>
<p>Next let’s generate a summary tree.</p>
<pre class="r"><code># generate a summary tree
trace = readTreeTrace(&quot;output/bears_FBD_morpho.trees&quot;)
mccTree(trace, file = &quot;output/bears_FBD_morpho.mcc.tre&quot;, positiveBranchLengths = TRUE)
q()</code></pre>
<p>This might take a while to run. Meanwhile, you could move on to the
<a href="#next">Next tasks</a>.</p>
</div>
<div id="output" class="section level2">
<h2>Evaluating the output</h2>
<p>As before, open your <code>.log</code> file in Tracer.<br />
It should look something like this.</p>
<p><img src="trace5.png" width="85%" style="padding:10px" /></p>
<blockquote>
<p>Explore the output. Can you identify the different parameters? Has
the analysis converged?</p>
</blockquote>
<blockquote>
<p>What is the age of the most recent common ancestor of living
bears?</p>
</blockquote>
<p>Once you have your summary tree, open it in FigTree and play around
with the settings. See if you can get something that looks like
this.</p>
<p><img src="mcc_bears_fbd.png" width="85%" style="padding:10px" /></p>
<blockquote>
<p>Which candiates on your tree do you think might be sampled
ancestors?</p>
</blockquote>
</div>
<div id="next" class="section level2">
<h2>Next tasks</h2>
<p>Some extra things to try. <strong>Make sure you rename the output
files so you don’t overwrite the files you generated above!</strong></p>
<div id="add-molecular-sequence-data" class="section level3">
<h3>Add molecular sequence data</h3>
<p>Try adding the molecular sequence alignment that complements this
dataset. You can download this from <a
href="data/bears_cytb.nex">here</a>. This is an alignment of the
mitochondrial gene cytochrome oxidase B.</p>
<p>There’s just a few extra steps we need to add when we read in the
data, at the beginning of <code>main.Rev</code>.</p>
<p>First add the line that reads in the sequence data and create a
constant variable <code>cytb</code>.</p>
<pre class="r"><code>taxa &lt;- readTaxonData(&quot;data/bears_taxa.tsv&quot;)
morpho &lt;- readDiscreteCharacterData(&quot;data/bears_morphology.nex&quot;)
cytb &lt;- readDiscreteCharacterData(&quot;data/bears_cytb.nex&quot;)</code></pre>
<p>Next we need to make sure the datasets are cross compatible because
we don’t have molecular sequence data for all species.</p>
<pre class="r"><code>cytb.addMissingTaxa( taxa )
morpho.addMissingTaxa( taxa )</code></pre>
<p>We need to add a clock model for the molecular partition. Let’s use a
relaxed clock model. We’ll use the popular uncorrelated lognormal rates
model, which uses independent rates across branches. See <a
href="https://revbayes.github.io/tutorials/clocks/">this tutorial</a>
for more details of different clock models.</p>
<pre class="r"><code>branch_rates_mean ~ dnExponential(10.0)
moves.append( mvScale(branch_rates_mean, lambda = 0.5, tune = true, weight = 3.0) )

for(i in 1:num_branches){
   brates_cytb[i] ~ dnExponential(branch_rates_mean)
   moves.append( mvScale(brates_cytb[i], lambda = 1, tune = true, weight = 2.0))
}</code></pre>
<p>Now you can go ahead and set up the GTR+G model as you did above and
in <a href="https://revbayes.github.io/tutorials/ctmc/">previous
exercises</a> (or whatever substitution model you like). Watch you don’t
overwrite any variables that are already used for your mophological
partition, e.g., <code>Q</code> or <code>branch_rates</code>.</p>
<p>Add your new scripts to <code>main.Rev</code>.</p>
<pre class="r"><code>source(&quot;scripts/clock_UEC_cytb.Rev&quot;)
source(&quot;scripts/GTR+G.Rev&quot;)</code></pre>
<p>Update all your output file names, so you don’t overwrite any of your
previous output!</p>
<pre class="r"><code># parameters printed to file
monitors.append( mnModel(filename = &quot;output/bears_FBD_morpho+cytb.log&quot;, printgen = 100) )
# trees printed to file
monitors.append( mnFile(filename = &quot;output/bears_FBD_morpho+cytb.trees&quot;, printgen = 100, tree) )

...

# generate a summary tree
trace = readTreeTrace(&quot;output/bears_FBD_morpho+cytb.trees&quot;)
mccTree(trace, file = &quot;output/bears_FBD_morpho+cytb.mcc.tre&quot;, positiveBranchLengths = TRUE)</code></pre>
<p>That’s it!</p>
<blockquote>
<p>What differences do you notice in your output?</p>
</blockquote>
</div>
<div id="running-the-analysis-under-the-fbd-model-only"
class="section level3">
<h3>Running the analysis under the FBD model only</h3>
<p>Try running the analysis “under the prior”. Simply add
<code>underPrior = TRUE</code> to the MCMC run function.</p>
<pre class="r"><code>mymcmc.run(generations = 10000, tuningInterval = 1000, underPrior = TRUE)</code></pre>
<p>Note it’s a bit funny to refer to this as “under the prior” because
the fossil ages are data but RevBayes doesn’t consider the FBD as part
of the likelihood it prints out, so you can still choose the regular
option for running the analysis under the prior to distinguish the
signal coming from the FBD/fossil age data vs. the FBD combined with the
sequence/morpho data.</p>
<blockquote>
<p>Compare you results to the posterior obtained using in the previous
steps. What information do we gain from the morphological / molecular
data?</p>
</blockquote>
</div>
<div id="accounting-for-fossil-age-uncertainty" class="section level3">
<h3>Accounting for fossil age uncertainty</h3>
<p>Since we basically never know the exact age of any fossils, we
typically need to account for age uncertainty. You can explore the
impact of age uncertainty by adding the following commands to your
<code>FBD_dating.Rev</code> script. You first need to download <a
href="data/bears_taxa_age_uncertainty.tsv">a new file</a> containing the
correct minimum and maximum bounds for the fossil ages.</p>
<p>The following loop goes through each fossil in our dataset and
assigns a uniform distribution between the minimum and maximum bounds of
the known interval of uncertainty.</p>
<pre class="r"><code>fossils = tree.getFossils()

for(i in 1:fossils.size())
{
    t[i] := tmrca(tree, clade(fossils[i]))

    a_i = fossils[i].getMinAge()
    b_i = fossils[i].getMaxAge()

    F[i] ~ dnUniform(t[i] - b_i, t[i] - a_i)
    F[i].clamp( 0 )
}</code></pre>
<blockquote>
<p>Compare your results to the posterior obtained in the previous steps.
Do the results change?</p>
</blockquote>
</div>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
